DBAgenda.php

```diff
@@ -0,0 +1,317 @@
+            $this->updateEvents($vCalendarFilename_to_update);
+        }
+        
+        //$this->removeDeletedEvents($ETags);
+    }
+    
+    // delete local events that have been deleted on the server
+    protected function removeDeletedEvents(array $ETags) {
+        $vCalendarFilenames = [];
+        foreach($ETags as $vCalendarFilename => $ETag) {
+            $vCalendarFilenames[] = basename($vCalendarFilename);
+        }
+        
+        foreach($this->localcache->getAllEventsNames() as $vCalendarFilename){
+            if(in_array($vCalendarFilename, $vCalendarFilenames)) {
+                $this->log->debug("No need to remove ". $vCalendarFilename);
+            } else {
+                $this->log->info("Need to remove ". $vCalendarFilename);
+                $this->localcache->deleteEvent($vCalendarFilename);
+            }
+        }
+    }
+    
+    protected function updateEvent(array $event) {
+        $vCalendar = \Sabre\VObject\Reader::read($event['vCalendarRaw']);
+        $datetime_begin = $vCalendar->VEVENT->DTSTART->getDateTime();
+
+        //$this->debug->log("deleting: $event[vCalendarFilename]");
+        
+        $query = $this->pdo->prepare("DELETE FROM `events_attendees` WHERE vCalendarFilename=:vCalendarFilename;");
+        $query->execute(array(
+            'vCalendarFilename' => $event['vCalendarFilename']
+        ));
+
+        $query = $this->pdo->prepare("DELETE FROM `events_categories` WHERE vCalendarFilename=:vCalendarFilename;");
+        $query->execute(array(
+            'vCalendarFilename' => $event['vCalendarFilename']
+        ));
+
+        
+        $number_volunteers_required = null;
+        if(isset($vCalendar->VEVENT->CATEGORIES)) {
+            foreach($vCalendar->VEVENT->CATEGORIES as $category) {
+                $category = (string)$category;
+                
+                if(is_null($number_volunteers_required) and !is_null($number_volunteers_required = is_number_of_attendee_category($category))) {
+                    continue;
+                }
+            }
+        }
+        
+        $query = $this->pdo->prepare("REPLACE INTO events (vCalendarFilename, ETag, datetime_begin, number_volunteers_required, vCalendarRaw) VALUES (:vCalendarFilename, :ETag, :datetime_begin, :number_volunteers_required, :vCalendarRaw)");
+        $query->execute(array(
+            'vCalendarFilename' =>  $event['vCalendarFilename'],
+            'ETag' => $event['ETag'],
+            'datetime_begin' => $datetime_begin->format('Y-m-d H:i:s'),
+            'number_volunteers_required' => $number_volunteers_required,
+            'vCalendarRaw' => $event['vCalendarRaw']
+        ));
+        
+        if(isset($vCalendar->VEVENT->ATTENDEE)) {
+            foreach($vCalendar->VEVENT->ATTENDEE as $attendee) {
+                $mail = str_replace("mailto:", "", (string)$attendee);
+
+                $query = $this->pdo->prepare("SELECT * FROM attendees WHERE email=:email;");
```

Comment:
potential optimization for later: retrieve all the users we need in a single call, before the loop (but let's not worry about that for the moment)